internal/gatewayapi/translator.go

```diff
@@ -315,28 +328,100 @@ func (t *Translator) Translate(resources *resource.Resources) (*TranslateResult,
 	acceptedGateways, failedGateways []*GatewayContext,
 ) {
+	envoyproxyMap := make(map[types.NamespacedName]*egv1a1.EnvoyProxy, len(resources.EnvoyProxiesForGateways)+1)
+	envoyproxyValidationErrorMap := make(map[types.NamespacedName]error, len(resources.EnvoyProxiesForGateways))
+
+	// if EnvoyProxy not found, provider layer set GC status to not accepted.
+	// if EnvoyProxy found but invalid, set GC status to not accepted,
+	// otherwise set GC status to accepted.
+	if ep := resources.EnvoyProxyForGatewayClass; ep != nil {
+		err := validateEnvoyProxy(ep)
+		if err != nil {
+			t.Logger.Error(err, "Skipping GatewayClass because EnvoyProxy is invalid",
+				"gatewayclass", t.GatewayClassName,
+				"envoyproxy", ep.Name, "namespace", ep.Namespace)
+			status.SetGatewayClassAccepted(resources.GatewayClass,
+				false, string(gwapiv1.GatewayClassReasonInvalidParameters),
+				fmt.Sprintf("%s: %v", status.MsgGatewayClassInvalidParams, err))
+			return
+		}
+
+		// TODO: remove this nil check after we update all the testdata.
+		if resources.GatewayClass != nil {
+			status.SetGatewayClassAccepted(
+				resources.GatewayClass,
+				true,
+				string(gwapiv1.GatewayClassReasonAccepted),
+				status.MsgValidGatewayClass)
+		}
+
+		key := utils.NamespacedName(ep)
+		envoyproxyMap[key] = ep
+		// we didn't append to envoyproxyValidatioErrorMap because it's valid.
+	}
+
+	for _, ep := range resources.EnvoyProxiesForGateways {
+		key := utils.NamespacedName(ep)
+		envoyproxyMap[key] = ep
+		if err := validateEnvoyProxy(ep); err != nil {
+			envoyproxyValidationErrorMap[key] = err
+		}
+	}
+
 	for _, gateway := range resources.Gateways {
 		if gateway == nil {
+			// Should not happen
 			panic("received nil gateway")
 		}
 
-		if gateway.Spec.GatewayClassName == t.GatewayClassName {
-			gc := &GatewayContext{
-				Gateway: gateway,
-			}
+		logKeysAndValues := []any{
+			"namespace", gateway.Namespace, "name", gateway.Name,
+		}
+		if gateway.Spec.GatewayClassName != t.GatewayClassName {
+			t.Logger.Info("Skipping Gateway because GatewayClassName doesn't match", logKeysAndValues...)
+			continue
+		}
+
+		gCtx := &GatewayContext{
+			Gateway: gateway,
+		}
 
-			// Gateways that are not accepted by the controller because they reference an invalid EnvoyProxy.
-			if status.GatewayNotAccepted(gc.Gateway) {
-				failedGateways = append(failedGateways, gc)
-			} else {
-				gc.ResetListeners(resources)
-				acceptedGateways = append(acceptedGateways, gc)
+		// Gateways that are not accepted by the controller because they reference an invalid EnvoyProxy.
+		if status.GatewayNotAccepted(gCtx.Gateway) {
+			failedGateways = append(failedGateways, gCtx)
+			t.Logger.Info("EnvoyProxy for Gateway not found ", logKeysAndValues...)
+			continue
+		}
+
+		gCtx.ResetListeners(resources, envoyproxyMap)
+		if ep := gCtx.envoyProxy; ep != nil {
+			key := utils.NamespacedName(ep)
+			if err, exits := envoyproxyValidationErrorMap[key]; exits {
```

Comment:
on L389
```
		// Gateways that are not accepted by the controller because they reference an invalid EnvoyProxy.
		if status.GatewayNotAccepted(gCtx.Gateway) {
```

but we are agin setting the status for the same reason, is this correct ?